tekton-listener/pkg/reconciler/tektonlistener/reconciler.go

```diff
@@ -0,0 +1,210 @@
+	kubeclientset kubernetes.Interface
+	// Listing cloud event listeners
+	tektonListenerLister listers.TektonListenerLister
+	// logger for inner info
+	logger *zap.SugaredLogger
+}
+
+// Check that we implement the controller.Reconciler interface.
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// NewController returns a new cloud events listener controller
+func NewController(
+	kubeclientset kubernetes.Interface,
+	tektonListenerInformer informers.TektonListenerInformer,
+) *controller.Impl {
+	// Enrich the logs with controller name
+	logger, _ := logging.NewLogger("", "tekton-listener")
+
+	r := &Reconciler{
+		kubeclientset:        kubeclientset,
+		tektonListenerLister: tektonListenerInformer.Lister(),
+		logger:               logger,
+	}
+	impl := controller.NewImpl(r, logger, "TektonListener",
+		reconciler.MustNewStatsReporter("TektonListener", r.logger))
+
+	logger.Info("Setting up tekton-listener event handler")
+	// Set up an event handler for when TektonListener resources change
+	tektonListenerInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    impl.Enqueue,
+		UpdateFunc: controller.PassNew(impl.Enqueue),
+	})
+
+	return impl
+}
+
+// Reconcile will create the necessary statefulset to manage the listener process
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	c.logger.Info("tekton-listener-reconcile")
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		c.logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	pl, err := c.tektonListenerLister.TektonListeners(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.Errorf("listener %q in work queue no longer exists", key)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if pl.Spec.PipelineRunSpec == (&pipelinev1alpha1.PipelineRunSpec{}) {
+		c.logger.Error("PipelineRunSpec must not be empty")
+		return nil
+	}
+
+	pl = pl.DeepCopy()
+	setName := pl.Name + "-statefulset"
+
+	containerEnv := []corev1.EnvVar{
+		{
+			Name:  "EVENT_TYPE",
+			Value: pl.Spec.EventType,
+		},
+		{
+			Name:  "EVENT",
+			Value: pl.Spec.Event,
+		},
+		{
+			Name:  "NAMESPACE",
+			Value: pl.Spec.Namespace,
+		},
+		{
+			Name:  "LISTENER_RESOURCE",
+			Value: pl.Name,
+		},
+		{
+			Name:  "SERVICEACCOUNT",
+			Value: pl.Spec.PipelineRunSpec.ServiceAccount,
+		},
+	}
+
+	c.logger.Infof("launching tekton-listener %s with type: %s namespace: %s",
+		pl.Name,
+		pl.Spec.EventType,
+		pl.Spec.Namespace,
+	)
+
+	// Create a stateful set for the listener. It mounts a secret containing the build information.
+	// The build spec may contain sensetive data and therefore the whole thing seems safest/easiest as a secret
+	set := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      setName,
+			Namespace: pl.Namespace,
+		},
+		Spec: appsv1.StatefulSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"statefulset": pl.Name + "-statefulset"},
```

Comment:
On a practical note: I was just trying out this code, and I think it would be helpful if I could specify labels in the EventBinding that will appear in the StatefulSet (and in turn for the listener pod that it creates). I was testing by creating an EventBinding, and at the moment I need to manually create a service for the listener pod. I think it would be helpful if I was able to specify my own labels for assigning the service. (Also, I don't think the `app: test-build-tekton-listener` selector in the `hack/listener.yaml` works at the moment, because the statefulset does not create any labels with `app: ...`)

On the other hand, it might be nice to create the listener service with the listener pod (stateful set) at some point in the future. Maybe an optional parameter in the TektonListener to specify the kind of service you would like to create.

Reply:
Thank you for bringing this topic up, you are totally right. As a first step, Ive made a change which passes labels down: `Eventbinding`->`Listener`->`StatefulSet`->`Pod`. This seems like an easy first step but I would love input.